mne/bem.py

```diff
@@ -12,10 +12,512 @@
+    return omega
+
+
+def _correct_auto_elements(surf, mat):
+    """Improve auto-element approximation..."""
+    pi2 = 2.0 * np.pi
+    tris_flat = surf['tris'].ravel()
+    misses = pi2 - mat.sum(axis=1)
+    for j, miss in enumerate(misses):
+        # How much is missing?
+        n_memb = len(surf['neighbor_tri'][j])
+        # The node itself receives one half
+        mat[j, j] = miss / 2.0
+        # The rest is divided evenly among the member nodes...
+        miss /= (4.0 * n_memb)
+        members = np.where(j == tris_flat)[0]
+        mods = members % 3
+        offsets = np.array([[1, 2], [-1, 1], [-1, -2]])
+        tri_1 = members + offsets[mods, 0]
+        tri_2 = members + offsets[mods, 1]
+        for t1, t2 in zip(tri_1, tri_2):
+            mat[j, tris_flat[t1]] += miss
+            mat[j, tris_flat[t2]] += miss
+    return
+
+
+def _fwd_bem_lin_pot_coeff(surfs):
+    """Calculate the coefficients for linear collocation approach"""
+    # taken from fwd_bem_linear_collocation.c
+    nps = [surf['np'] for surf in surfs]
+    np_tot = sum(nps)
+    coeff = np.zeros((np_tot, np_tot))
+    offsets = np.cumsum(np.concatenate(([0], nps)))
+    for si_1, surf1 in enumerate(surfs):
+        rr_ord = np.arange(nps[si_1])
+        for si_2, surf2 in enumerate(surfs):
+            logger.info("        %s (%d) -> %s (%d) ..." %
+                        (_bem_explain_surface(surf1['id']), nps[si_1],
+                         _bem_explain_surface(surf2['id']), nps[si_2]))
+            tri_rr = surf2['rr'][surf2['tris']]
+            tri_nn = surf2['tri_nn']
+            tri_area = surf2['tri_area']
+            submat = coeff[offsets[si_1]:offsets[si_1 + 1],
+                           offsets[si_2]:offsets[si_2 + 1]]  # view
+            for k in range(surf2['ntri']):
+                tri = surf2['tris'][k]
+                if si_1 == si_2:
+                    skip_idx = ((rr_ord == tri[0]) |
+                                (rr_ord == tri[1]) |
+                                (rr_ord == tri[2]))
+                else:
+                    skip_idx = []
+                # No contribution from a triangle that
+                # this vertex belongs to
+                # if sidx1 == sidx2 and (tri == j).any():
+                #     continue
+                # Otherwise do the hard job
+                coeffs = _lin_pot_coeff(surf1['rr'], tri_rr[k], tri_nn[k],
+                                        tri_area[k])
+                coeffs[skip_idx] = 0.
+                submat[:, tri] -= coeffs
+            if si_1 == si_2:
+                _correct_auto_elements(surf1, submat)
+    return coeff
+
+
+def _fwd_bem_multi_solution(solids, gamma, nps):
+    """Do multi surface solution
+
+      * Invert I - solids/(2*M_PI)
+      * Take deflation into account
+      * The matrix is destroyed after inversion
+      * This is the general multilayer case
+
+    """
+    from scipy import linalg
```

Comment:
we nest certain scipy imports now to save on import speed, see `mne/tests/test_import_nesting.py`